File: application/backend/business/menu-v2/src/main/java/com/demo/poc/commons/core/interceptor/error/ResponseErrorSelector.java
package com.demo.poc.commons.core.interceptor.error;

import com.demo.poc.commons.core.errors.dto.ErrorDTO;
import com.demo.poc.commons.core.errors.dto.ErrorType;
import com.demo.poc.commons.custom.exceptions.ErrorDictionary;
import com.demo.poc.commons.core.errors.exceptions.GenericException;
import com.demo.poc.commons.core.properties.ConfigurationBaseProperties;
import com.demo.poc.commons.core.properties.ProjectType;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.ConstraintViolationException;

import java.util.Optional;
import java.util.stream.Collectors;

import static com.demo.poc.commons.core.errors.dto.ErrorDTO.CODE_DEFAULT;

public class ResponseErrorSelector {

  public static <T extends Throwable> ErrorDTO toErrorDTO(ConfigurationBaseProperties properties, T exception) {
    ErrorDTO error = extractError(exception);
    String selectedCode = selectCustomCode(properties, error);
    error.setCode(selectedCode);

    String selectedMessage = selectMessage(properties, error);
    error.setMessage(selectedMessage);

    return error;
  }

  private static String selectCustomCode(ConfigurationBaseProperties properties, ErrorDTO error) {
    ProjectType projectType = selectProjectType(properties);

    return ProjectType.BFF.equals(projectType)
        ? CODE_DEFAULT
        : Optional.ofNullable(error)
        .map(ErrorDTO::getCode)
        .orElseGet(() -> CODE_DEFAULT);
  }

  private static String selectMessage(ConfigurationBaseProperties properties, ErrorDTO error) {
    String defaultMessage = ErrorDTO.getDefaultError(properties).getMessage();
    ProjectType projectType = selectProjectType(properties);

    if(ProjectType.BFF.equals(projectType))
      return defaultMessage;

    return Optional.ofNullable(ErrorDTO.getMatchMessage(properties, error.getCode()))
        .orElseGet(() -> Optional.ofNullable(error.getMessage()).orElse(defaultMessage));
  }

  private static ProjectType selectProjectType(ConfigurationBaseProperties properties) {
    return ProjectType.valueOf(properties.projectType().orElse(ProjectType.MS.name()));
  }

  private static <T extends Throwable> ErrorDTO extractError(T exception) {
    ErrorDTO error = null;

    if (exception instanceof GenericException genericException)
      error = genericException.getErrorDetail();

    if(exception instanceof ConstraintViolationException constraintViolationException)
      error = extractError(constraintViolationException);

    return error;
  }

  private static ErrorDTO extractError(ConstraintViolationException exception) {
    String message = exception
        .getConstraintViolations()
        .stream()
        .map(ConstraintViolation::getMessage)
        .collect(Collectors.joining(";"));

    return ErrorDTO.builder()
        .code(ErrorDictionary.INVALID_FIELD.getCode())
        .message(message)
        .type(ErrorType.BUSINESS)
        .build();
  }
}

File: backend/microservices/product-v1/commons/core/errors/selector/response_error_selector.go
package selector

import (
	"poc/commons/core/constants"
	errorDto "poc/commons/core/errors/dto"
	"poc/commons/custom/properties"
)

type ResponseErrorSelector struct{}

func NewResponseErrorSelector() *ResponseErrorSelector {
	return &ResponseErrorSelector{}
}

func (selector *ResponseErrorSelector) ToErrorDto(err error) errorDto.ErrorDto {
	baseError := extractError(err)
	projectType := selector.selectProjectType()
	baseError.Code = selectCustomCode(baseError.Code, projectType)
	baseError.Message = selector.selectMessage(baseError.Message, baseError.Code, projectType)
	return baseError
}

func extractError(err error) errorDto.ErrorDto {
	if genericError, isGenericError := err.(errorDto.GenericError); isGenericError {
		recoveredError := errorDto.ErrorDto{
			Origin: string(genericError.Origin),
			Code:   genericError.Code,
		}
		if message := genericError.Message; message != constants.EMPTY {
			recoveredError.Message = message
		}
		return recoveredError
	}
	return errorDto.ErrorDto{Origin: string(errorDto.ERROR_ORIGIN_OWN)}
}

func (selector *ResponseErrorSelector) selectProjectType() properties.ProjectType {
	if projectType := properties.ProjectType(properties.Properties.ProjectType); projectType != constants.EMPTY {
		return projectType
	}
	return properties.PROJECT_TYPE_MS
}

func selectCustomCode(code string, projectType properties.ProjectType) string {
	if projectType == properties.PROJECT_TYPE_BFF {
		return errorDto.CODE_DEFAULT
	}
	if code == constants.EMPTY {
		return errorDto.CODE_DEFAULT
	}
	return code
}

func (selector *ResponseErrorSelector) selectMessage(originalMessage, code string, projectType properties.ProjectType) string {
	defaultMessage := properties.Properties.ErrorMessages[errorDto.CODE_DEFAULT]

	if projectType == properties.PROJECT_TYPE_BFF {
		return defaultMessage
	}
	if message, existsMessage := properties.Properties.ErrorMessages[code]; existsMessage {
		return message
	}
	if originalMessage != constants.EMPTY {
		return originalMessage
	}
	return defaultMessage
}
